libs/platform/src/lib/table-helpers/directives/table-virtual-scroll.directive.ts

```diff
@@ -79,62 +135,174 @@ export class TableVirtualScrollDirective extends TableVirtualScroll implements O
      * Calculates rows in viewport.
      */
     calculateVirtualScrollRows(): void {
-        if (!this.virtualScroll || !this.bodyHeight) {
-            return;
-        }
+        if (this.scrollWholeRows) {
+            if (!this._focusableGridSubscription) {
+                this._focusableGridSubscription = this._table._focusableGrid.itemFocused.subscribe((event) => {
+                    this._focusedCell = event;
+                });
+            }
+            if (!this.virtualScroll || !this.bodyHeight) {
+                return;
+            }
 
-        const rowHeight = this.rowHeight + 1;
+            let startNodeIndex = 0;
+            const scrollTop = this._table.getTableState().scrollTopPosition;
+            if (scrollTop !== 0) {
+                startNodeIndex = Math.floor(scrollTop / this.rowHeight);
+            }
 
-        const rowsVisible = this._table._tableRowsVisible;
-        const rowsInViewPort = this._table.getRowsInViewport();
-        const totalNodeCount = rowsVisible.length;
-        const scrollTop = this._table.tableScrollable.getScrollTop();
+            const rowsVisible = this._table._tableRowsVisible;
+            const totalNodeCount = rowsVisible.length;
+            let visibleNodeCount =
+                Math.ceil(this._table.tableContainer.nativeElement.clientHeight / this.rowHeight) * this.renderAhead;
+            visibleNodeCount = Math.min(totalNodeCount - startNodeIndex, visibleNodeCount);
+            this.virtualScrollTotalHeight = totalNodeCount * this.rowHeight - visibleNodeCount * this.rowHeight;
 
-        let startNodeIndex = Math.floor(scrollTop / rowHeight) - this.renderAhead;
-        startNodeIndex = Math.max(0, startNodeIndex);
+            this._setRows(startNodeIndex);
 
-        let visibleNodeCount =
-            Math.ceil(this._table.tableContainer.nativeElement.clientHeight / rowHeight) + 2 * this.renderAhead;
-        visibleNodeCount = Math.min(totalNodeCount - startNodeIndex, visibleNodeCount);
+            this._table.tableScrollMockContainer.nativeElement.style.width = '1rem';
+            this._table.tableScrollMockContainer.nativeElement.style.maxHeight = this.bodyHeight;
```

Comment:
This is duplicated DOM query. This should be locally cached.
Ex: `const scrollMockContainer = this._table.tableScrollMockContainer.nativeElement`.